crates/bevy_render/src/lib.rs

```diff
@@ -144,196 +144,202 @@ impl Plugin for RenderPlugin {
-                )
-                .add_stage(RenderStage::Cleanup, SystemStage::parallel())
-                .init_resource::<RenderGraph>()
-                .insert_resource(RenderInstance(instance))
-                .insert_resource(device)
-                .insert_resource(queue)
-                .insert_resource(render_adapter)
-                .insert_resource(available_texture_formats)
-                .insert_resource(texture_format)
-                .insert_resource(adapter_info)
-                .insert_resource(pipeline_cache)
-                .insert_resource(asset_server);
-
-            let (sender, receiver) = bevy_time::create_time_channels();
-            app.insert_resource(receiver);
-            render_app.insert_resource(sender);
-
-            app.add_sub_app(RenderApp, render_app, move |app_world, render_app| {
-                #[cfg(feature = "trace")]
-                let _render_span = bevy_utils::tracing::info_span!("renderer subapp").entered();
-                {
-                    #[cfg(feature = "trace")]
-                    let _stage_span =
-                        bevy_utils::tracing::info_span!("stage", name = "reserve_and_flush")
-                            .entered();
-
-                    // reserve all existing app entities for use in render_app
-                    // they can only be spawned using `get_or_spawn()`
-                    let meta_len = app_world.entities().meta_len();
-                    render_app
-                        .world
-                        .entities()
-                        .reserve_entities(meta_len as u32);
-
-                    // flushing as "invalid" ensures that app world entities aren't added as "empty archetype" entities by default
-                    // these entities cannot be accessed without spawning directly onto them
-                    // this _only_ works as expected because clear_entities() is called at the end of every frame.
-                    unsafe { render_app.world.entities_mut() }.flush_as_invalid();
-                }
-
-                {
-                    #[cfg(feature = "trace")]
-                    let _stage_span =
-                        bevy_utils::tracing::info_span!("stage", name = "extract").entered();
-
-                    // extract
-                    extract(app_world, render_app);
-                }
-
-                {
-                    #[cfg(feature = "trace")]
-                    let _stage_span =
-                        bevy_utils::tracing::info_span!("stage", name = "prepare").entered();
-
-                    // prepare
-                    let prepare = render_app
-                        .schedule
-                        .get_stage_mut::<SystemStage>(RenderStage::Prepare)
-                        .unwrap();
-                    prepare.run(&mut render_app.world);
-                }
-
-                {
+            if let Some(window) = windows.get_primary() {
```

Comment:
Why prevent building the render plugin if there is no primary window? I think the previous behavior was intentional. There might be applications, that do not have a graphical interface, but use e.g. compute shaders.

Reply:
Got it.  Needed to move that if inside the block for `let surface = {`.